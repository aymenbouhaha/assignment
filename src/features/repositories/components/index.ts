export * from "./primary-language";
export * from "./repository-item";
export * from "./repository-list";
export * from "./repository-page-header";
export * from "./repository-search-filter-bar";
export * from "./license";
