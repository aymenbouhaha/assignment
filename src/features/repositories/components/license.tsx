import { Icon } from "@shared/components/icons/icons.tsx";

export const License = ({ name }: { name: string }) => {
	return (
		<div className="flex gap-1.5 items-center">
			<Icon.Balance className={"size-6"} />
			<div className="text-p3 text-primary-green">{name}</div>
		</div>
	);
};
