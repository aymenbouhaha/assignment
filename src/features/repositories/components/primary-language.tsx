import { Icon } from "@shared/components/icons/icons.tsx";
import React from "react";

type PrimaryLanguageProps = { name: string; color: string };

/**
 * The `PrimaryLanguage` component is a reusable UI element designed to display a programming language or primary language name alongside a colored dot icon.
 *
 * @component
 * @example
 * ```tsx
 * <PrimaryLanguage name="JavaScript" color="#F7DF1E" />
 * ```
 *
 * @param {Object} props - Component props.
 * @param {string} props.name - The name of the language to be displayed (e.g., "JavaScript", "Python").
 * @param {string} props.color - A color string (e.g., hex or RGB) used for the dot icon and text.
 *
 * @returns {React.JSX} The rendered `PrimaryLanguage` component.
 */
export const PrimaryLanguage = ({ name, color }: PrimaryLanguageProps): React.JSX.Element => {
	return (
		<div className="flex gap-1.5 items-center">
			<Icon.Dot
				className={"size-3.5 lg:size-2.5"}
				style={{
					fill: color,
				}}
			/>
			<div
				className="text-p3 lg:text-p4"
				style={{
					color: color,
				}}
			>
				{name}
			</div>
		</div>
	);
};
