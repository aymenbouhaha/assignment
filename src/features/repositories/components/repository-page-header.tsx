import { Icon } from "@shared/components/icons/icons.tsx";

export const RepositoryPageHeader = () => {
	return (
		<div className="flex gap-3 items-center ">
			<Icon.UserGraphic className={"size-[80px]"} />
			<div className="flex flex-col justify-between h-full">
				<div className="text-h1 text-primary-black">Welcome back, Aymen Bouhaha! ğŸ‘‹</div>
				<div className="text-p21 text-grey-2">What are you up to today?</div>
			</div>
		</div>
	);
};
