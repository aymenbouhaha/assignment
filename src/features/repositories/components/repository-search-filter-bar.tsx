import { FormProviderWrapper } from "@shared/components/inputs/form-provider-wrapper.tsx";
import { useForm } from "react-hook-form";
import { ControlledTextInput } from "@shared/components/inputs/controlled-text-input.tsx";
import { ControlledSelectInput } from "@shared/components/inputs/controlled-select-input.tsx";
import {
	repositoryFilterDefault,
	RepositoryFilterModel,
	RepositoryFilterSchema,
} from "@/features/repositories/models/repository-filter.model.ts";
import { zodResolver } from "@hookform/resolvers/zod";

export const RepositorySearchFilterBar = () => {
	const form = useForm<RepositoryFilterModel>({
		resolver: zodResolver(RepositoryFilterSchema),
		defaultValues: repositoryFilterDefault,
	});
	return (
		<FormProviderWrapper form={form} className="flex gap-3" onSubmit={() => {}}>
			<div className="flex-[3]">
				<ControlledTextInput
					placeholder={"Type repository name"}
					fieldName={"repository"}
					leadIcon={{ icon: "Search" }}
				/>
			</div>
			<div className="flex-1">
				<ControlledSelectInput
					fieldName={"language"}
					values={[{ placeholder: "Aymen", value: "Aymen" }]}
					placeholder={"Languages"}
				/>
			</div>
		</FormProviderWrapper>
	);
};
