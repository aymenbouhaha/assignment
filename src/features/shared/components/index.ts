export * from "./ui";
export * from "./loaders";
export * from "./others";
export * from "./dropdown/dropdown";
export * from "./buttons/pagination-button";
export * from "./icons/icons";
