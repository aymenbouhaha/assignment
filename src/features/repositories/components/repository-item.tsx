import { PrimaryLanguage } from "./primary-language.tsx";
import { License } from "@/features/repositories/components/license.tsx";
import { RepositoryModel } from "@/features/repositories/models/repository.model.ts";
import React from "react";

/**
 * The `RepositoryItem` component displays information about a repository, including its name, description, primary language, and license.
 *
 * @component
 * @example
 * ```tsx
 * <RepositoryItem
 *   repository={{
 *     name: "MyRepo",
 *     url: "https://github.com/user/myrepo",
 *     isPrivate: false,
 *     description: "A sample repository.",
 *     primaryLanguage: { name: "TypeScript", color: "#3178C6" },
 *     licenseInfo: { name: "MIT" },
 *     parent: {
 *       url: "https://github.com/parent/repo",
 *       nameWithOwner: "parent/repo",
 *     },
 *   }}
 * />
 * ```
 *
 * @param {Object} props - Component props.
 * @param {RepositoryModel} props.repository - The repository data model containing details to display.
 *
 * @returns {React.JSX} The rendered `RepositoryItem` component.
 */
export const RepositoryItem = ({ repository }: { repository: RepositoryModel }): React.JSX.Element => {
	return (
		<div className="border border-grey-border rounded-md py-2.5 px-3 flex justify-between">
			<div className="flex flex-col gap-2">
				<div className="flex-flex-col gap-1">
					<div className="flex gap-2 items-center flex-wrap">
						<a
							className="text-p1 !font-semibold text-primary-purple hover:underline hover:underline-offset-3 lg:text-p2 "
							href={repository.url}
						>
							{repository.name}
						</a>
						<div className="px-1 py-[2px] border border-grey-border text-grey-1 text-p4 lg:text-p6 rounded">
							{repository.isPrivate ? "Private" : "Public"}
						</div>
					</div>
					{repository.parent && (
						<div className="text-p4 text-grey-1 lg:text-p5">
							Forked from{" "}
							<a href={repository.parent.url} className={"underline underline-offset-3"}>
								{repository.parent.nameWithOwner}
							</a>
						</div>
					)}
				</div>
				<div className="text-p3 text-primary-black lg:text-p5">{repository.description}</div>
				<div className="flex gap-3">
					{repository.primaryLanguage && (
						<PrimaryLanguage name={repository.primaryLanguage.name} color={repository.primaryLanguage.color} />
					)}
					{repository.licenseInfo && <License name={repository.licenseInfo.name} />}
				</div>
			</div>
		</div>
	);
};
