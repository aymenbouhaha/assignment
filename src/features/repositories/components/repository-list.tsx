import { Skeleton } from "@shared/components/ui";
import { RepositoryItem } from "@/features/repositories/components/repository-item.tsx";
import { RepositoryModel } from "@/features/repositories/models/repository.model.ts";
import { cn } from "@/lib/utils.ts";

export const RepositoryList = ({
	loading,
	repositories,
	owner,
	containerClassName,
}: {
	owner?: string;
	repositories: RepositoryModel[];
	loading?: boolean;
	containerClassName?: string;
}) => {
	return (
		<>
			<div
				className={cn(
					"flex flex-col gap-3 border border-grey-border bg-[#fff] p-3 rounded-[10px]",
					containerClassName,
					{
						"h-80 justify-center items-center": !loading && repositories.length === 0,
					},
				)}
			>
				{loading ? (
					[0, 1, 2, 3, 5, 6].map((item) => <Skeleton key={`repository-skeleton.${item}`} className={"w-full h-40"} />)
				) : repositories.length ? (
					repositories.map((item, index) => (
						<RepositoryItem key={`repository-item.${item.id}.${index}`} repository={item} />
					))
				) : (
					<div className="text-grey-1 text-h1 !font-bold lg:text-p2 text-center">
						{owner ?? "User"} doesnâ€™t have any repositories that match.
					</div>
				)}
			</div>
		</>
	);
};
