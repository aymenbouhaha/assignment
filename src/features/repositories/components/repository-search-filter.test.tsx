import { render, screen, fireEvent } from "@testing-library/react";
import { RepositorySearchFilterBar } from "@repositories/components";
import { repositoryFilterDefault } from "@repositories/models/repository-filter.model.ts";
import { describe, it, expect, vi, beforeEach } from "vitest";

describe("RepositorySearchFilterBar", () => {
	const mockOnFilterValuesChange = vi.fn();

	beforeEach(() => {
		mockOnFilterValuesChange.mockClear();
	});

	it("renders input fields correctly when languages are provided", () => {
		const { container } = render(
			<RepositorySearchFilterBar
				languages={["JavaScript", "Python", "Java"]}
				onFilterValuesChange={mockOnFilterValuesChange}
			/>,
		);

		expect(container).toMatchSnapshot();
	});

	it("displays skeleton loaders when languages are not provided", () => {
		const { container } = render(<RepositorySearchFilterBar onFilterValuesChange={mockOnFilterValuesChange} />);

		expect(container).toMatchSnapshot();
	});

	it("calls onFilterValuesChange with updated repository name", () => {
		render(
			<RepositorySearchFilterBar
				languages={["JavaScript", "Python"]}
				onFilterValuesChange={mockOnFilterValuesChange}
			/>,
		);

		const repoInput = screen.getByPlaceholderText("Type repository name");
		fireEvent.change(repoInput, { target: { value: "my-repo" } });

		expect(mockOnFilterValuesChange).toHaveBeenCalledWith({
			...repositoryFilterDefault,
			repository: "my-repo",
		});
	});

	it("calls onFilterValuesChange with updated language", () => {
		render(
			<RepositorySearchFilterBar
				languages={["JavaScript", "Python"]}
				onFilterValuesChange={mockOnFilterValuesChange}
			/>,
		);

		const languageSelectTrigger = screen.getByText("Languages");
		expect(languageSelectTrigger).toBeInTheDocument();

		fireEvent.click(languageSelectTrigger!);

		const pythonOption = screen.getByText("Python");
		expect(pythonOption).toBeInTheDocument();
		fireEvent.click(pythonOption!);

		expect(mockOnFilterValuesChange).toHaveBeenCalledWith({
			...repositoryFilterDefault,
			language: "Python",
		});
	});

	it("updates both repository name and language correctly", () => {
		render(
			<RepositorySearchFilterBar
				languages={["JavaScript", "Python"]}
				onFilterValuesChange={mockOnFilterValuesChange}
			/>,
		);

		const repoInput = screen.getByPlaceholderText("Type repository name");
		fireEvent.change(repoInput, { target: { value: "new-repo" } });

		const languageSelectTrigger = screen.getByText("Languages");
		expect(languageSelectTrigger).toBeInTheDocument();

		fireEvent.click(languageSelectTrigger!);

		const pythonOption = screen.getByText("JavaScript");
		expect(pythonOption).toBeInTheDocument();
		fireEvent.click(pythonOption!);

		expect(mockOnFilterValuesChange).toHaveBeenLastCalledWith({
			repository: "new-repo",
			language: "JavaScript",
		});
	});
});
