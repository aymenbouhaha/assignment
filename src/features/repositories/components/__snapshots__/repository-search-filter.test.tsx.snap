// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RepositorySearchFilterBar > displays skeleton loaders when languages are not provided 1`] = `
<div>
  <div
    class="flex gap-3 lg:flex-col lg:gap-2"
  >
    <div
      class="flex-[3]"
    >
      <div
        class="animate-pulse rounded-md bg-grey-border h-12"
        data-testid="skeleton-loader"
      />
    </div>
    <div
      class="flex-1"
    >
      <div
        class="animate-pulse rounded-md bg-grey-border h-12"
        data-testid="skeleton-loader"
      />
    </div>
  </div>
</div>
`;

exports[`RepositorySearchFilterBar > renders input fields correctly when languages are provided 1`] = `
<div>
  <div
    class="flex gap-3 lg:flex-col lg:gap-2"
  >
    <div
      class="flex-[3]"
    >
      <div
        class="w-full relative"
      >
        <div
          class="absolute inset-y-0 left-3 flex items-center"
        >
          <svg
            class="size-6"
            fill="none"
            height="18"
            viewBox="0 0 18 18"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10.7888 11.5843C8.8011 13.1763 5.89152 13.051 4.04893 11.2084C2.0719 9.23138 2.0719 6.02598 4.04893 4.04895C6.02596 2.07192 9.23136 2.07192 11.2084 4.04895C13.051 5.89154 13.1763 8.8011 11.5843 10.7888L15.451 14.6555C15.6707 14.8752 15.6707 15.2314 15.451 15.451C15.2313 15.6707 14.8752 15.6707 14.6555 15.451L10.7888 11.5843ZM4.84443 10.4129C3.30674 8.87522 3.30674 6.38213 4.84443 4.84444C6.38212 3.30675 8.87521 3.30676 10.4129 4.84444C11.9495 6.38101 11.9506 8.87156 10.4163 10.4095C10.4151 10.4106 10.414 10.4118 10.4129 10.4129C10.4118 10.414 10.4106 10.4152 10.4095 10.4163C8.87155 11.9506 6.38099 11.9495 4.84443 10.4129Z"
              fill="#72767C"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <input
          class="file:bg-transparent file:text-sm file:text-zinc-950 flex w-full rounded-md border border-grey-border px-3 py-2.5 text-p21 lg:text-p3 ring-0 outline-none focus:border-grey-2 file:border-0 file:font-medium placeholder:text-grey-1 disabled:cursor-not-allowed disabled:opacity-50 pl-10"
          placeholder="Type repository name"
          value=""
        />
      </div>
    </div>
    <div
      class="flex-1"
    >
      <div
        class="w-full h-full"
      >
        <div
          class="relative w-full"
        >
          <button
            aria-autocomplete="none"
            aria-controls="radix-:r0:"
            aria-expanded="false"
            class="flex w-full items-center justify-between whitespace-nowrap rounded-md border border-grey-border focus:border-grey-2 bg-[#fff] px-3 py-2.5 ring-0 !text-p21 lg:!text-p3 outline-none disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 text-grey-1"
            data-placeholder=""
            data-state="closed"
            dir="ltr"
            role="combobox"
            type="button"
          >
            <span
              style="pointer-events: none;"
            >
              Languages
            </span>
          </button>
          <svg
            class="lucide lucide-chevron-down stroke-grey-2 absolute size-6 top-1/2 -translate-y-1/2 right-3 cursor-pointer"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
